File: 03-week-3--additional-practice/04-bonus-5--consecutive-substrings/ruby/consecutive_substrings.rb
def consecutive_substrings(string)
  # type your code in here
  sub_strings = []
  char_list = string.split('')
  char_list.each_with_index do |char, index|
    sub_strings << char
    letter = char

    string[index + 1..-1].split('').each_with_index do |next_letter|
      letter += next_letter
      sub_strings << letter
    end
  end

  sub_strings
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']"
  print "=> " 
  print consecutive_substrings('abc')

  puts

  puts "Expecting: ['a']"
  print "=> " 
  print consecutive_substrings('a')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 03-week-3--additional-practice/04-bonus-5--consecutive-substrings/javascript/consecutive_substrings.js
function consecutiveSubstrings(string) {
  // type your code here
  if (string.length <= 1) {return string.split('')}
  const characters = string.split('')
  const numberOfSubstrings = ({num, count}) => {return num > 0 ? numberOfSubstrings({num: num - 1, count: count + num}) : count}
  const subCount = numberOfSubstrings({num: characters.length, count: 0})

  const populateSubstringArr = ({subArr, count, nextCount, res}) => {
    if (subArr.length < subCount) {
      const currentElement = res ? characters[count] : subArr[subArr.length - 1] + characters[count]
      if (count === characters.length - 1) {
        return populateSubstringArr({subArr: subArr.concat([currentElement]), count: nextCount, nextCount: nextCount + 1, res: true})
      } else {
        return populateSubstringArr({subArr: subArr.concat([currentElement]), count: count + 1, nextCount: nextCount, res: false})
      }
    } else {
      return subArr
    }
  }
  return populateSubstringArr({subArr: [], count: 0, nextCount: 1, res: true})
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']");
  console.log("=>", consecutiveSubstrings('abc'));

  console.log("");

  console.log("Expecting: ['a']");
  console.log("=>", consecutiveSubstrings('a'));
}

module.exports = consecutiveSubstrings;

// Please add your pseudocode to this file
// And a written explanation of your solution
